cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java

```diff
@@ -0,0 +1,287 @@
+      // don't count this node
+      sinksAndReduceNodes = Sets.difference(sinksAndReduceNodes, ImmutableSet.of(node));
+
+      if (sinksAndReduceNodes.size() > 1) {
+        for (String reduceNodeConnector : Sets.intersection(sinksAndReduceNodes, reduceNodes)) {
+          addConnectorInFrontOf(reduceNodeConnector, addedAlready);
+        }
+      }
+    }
+
+    /*
+        Find nodes that have input from multiple sources and add them to the connectors set.
+        We can probably remove this part once we support multiple sources. Even though we don't support
+        multiple sources today, the fact that we support forks means we have to deal with the multi-input case
+        and break it down into separate phases. For example:
+
+                    |---> reduce1 ---|
+          source ---|                |---> sink
+                    |---> reduce2 ---|
+
+        From the previous section, both reduces will get a connector inserted in front:
+
+                    |---> reduce1.connector               reduce1.connector ---> reduce1 ---|
+          source ---|                              =>                                       |---> sink
+                    |---> reduce2.connector               reduce2.connector ---> reduce2 ---|
+
+        Since we don't support multi-input yet, we need to convert that further into 3 phases:
+
+          connector1 ---> reduce1 ---> sink.connector
+                                                            =>       sink.connector ---> sink
+          connector2 ---> reduce2 ---> sink.connector
+
+        To find these nodes, we traverse the graph in order and keep track of sources that have a path to each node
+        with a map of node -> [ sources that have a path to the node ]
+        if we find that a node is accessible by more than one source, we insert a connector in front of it and
+        reset all sources for that node to its connector
+     */
+    SetMultimap<String, String> nodeSources = HashMultimap.create();
+    for (String source : sources) {
+      nodeSources.put(source, source);
+    }
+    for (String connector : connectors) {
+      nodeSources.put(connector, connector);
+    }
+    for (String node : linearize()) {
+      if (sinks.contains(node)) {
+        continue;
+      }
+      Set<String> connectedSources = nodeSources.get(node);
+      // if more than one source is connected to this node, then this node is a connector.
+      // keep track of it, and wipe its node sources to just contain itself
+      if (connectedSources.size() > 1) {
+        addConnectorInFrontOf(node, addedAlready);
+        connectedSources = new HashSet<>();
+        connectedSources.add(node);
+        nodeSources.replaceValues(node, connectedSources);
+      }
+      for (String nodeOutput : getNodeOutputs(node)) {
+        // don't propagate to connectors, since they are counted as source and replace any inputs they may have
+        if (connectors.contains(nodeOutput)) {
```

Comment:
ok i remember why it is needed now. If my output is a connector, I don't want to add any sources to the connector because the connector itself counts as a source. Without this, I can add a source to the connector, and when we get to the connector it will have multiple inputs coming in and we'll insert a connector in front of a connector.

Instead of this, i think this logic is made clearer if i just clear out the inputs to the node if the node is a connector.